{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **GPU cehck**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-01-10 07:21:30.560963: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcudart.so.10.1\n",
      "2022-01-10 07:21:52.805566: I tensorflow/compiler/jit/xla_cpu_device.cc:41] Not creating XLA devices, tf_xla_enable_xla_devices not set\n",
      "2022-01-10 07:21:52.807943: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcuda.so.1\n",
      "2022-01-10 07:21:52.861931: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1720] Found device 0 with properties: \n",
      "pciBusID: 0000:83:00.0 name: GeForce RTX 2080 Ti computeCapability: 7.5\n",
      "coreClock: 1.665GHz coreCount: 68 deviceMemorySize: 10.76GiB deviceMemoryBandwidth: 573.69GiB/s\n",
      "2022-01-10 07:21:52.861976: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcudart.so.10.1\n",
      "2022-01-10 07:21:53.099155: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcublas.so.10\n",
      "2022-01-10 07:21:53.099278: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcublasLt.so.10\n",
      "2022-01-10 07:21:53.294322: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcufft."
     ]
    },
    {
     "data": {
      "text/plain": [
       "OrderedDict([('cpu_compiler',\n",
       "              '/home/builder/ktietz/aggregate/tensorflow_recipes/ci_cpu/tensorflow-base_1614534878484/_build_env/bin/x86_64-conda_cos6-linux-gnu-gcc'),\n",
       "             ('cuda_compute_capabilities',\n",
       "              ['compute_35',\n",
       "               'compute_52',\n",
       "               'compute_60',\n",
       "               'compute_61',\n",
       "               'compute_70',\n",
       "               'compute_75']),\n",
       "             ('cuda_version', '10.1'),\n",
       "             ('cudnn_version', '7'),\n",
       "             ('is_cuda_build', True),\n",
       "             ('is_rocm_build', False)])"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "so.10\n",
      "2022-01-10 07:21:53.505071: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcurand.so.10\n",
      "2022-01-10 07:21:53.674582: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcusolver.so.10\n",
      "2022-01-10 07:21:53.817902: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcusparse.so.10\n",
      "2022-01-10 07:21:54.182371: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcudnn.so.7\n",
      "2022-01-10 07:21:54.184811: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1862] Adding visible gpu devices: 0\n"
     ]
    }
   ],
   "source": [
    "import tensorflow as tf\n",
    "tf.test.is_built_with_cuda()\n",
    "tf.config.list_physical_devices('GPU')\n",
    "tf.sysconfig.get_build_info()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **Module**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [],
   "source": [
    "from datetime import datetime\n",
    "import os\n",
    "import re\n",
    "import gc\n",
    "import pickle\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "from functools import reduce\n",
    "\n",
    "from sklearn.manifold import Isomap\n",
    "from sklearn.decomposition import PCA\n",
    "from sklearn.decomposition import KernelPCA\n",
    "from sklearn.cluster import KMeans\n",
    "from sklearn.impute import KNNImputer\n",
    "from sklearn.manifold import TSNE\n",
    "from sklearn.preprocessing import MinMaxScaler\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.cluster import DBSCAN\n",
    "import matplotlib.pyplot as plt\n",
    "import umap.umap_ as umap\n",
    "\n",
    "# keras\n",
    "from tensorflow.keras.models import Model\n",
    "from tensorflow.keras.layers import Dense, Input\n",
    "from tensorflow.keras.preprocessing import sequence\n",
    "from tensorflow.keras.callbacks import EarlyStopping"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **UDF**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def cancer_select(cols, cancer_type):\n",
    "    # phenotype\n",
    "    phe1 = pd.read_csv(\"https://gdc-hub.s3.us-east-1.amazonaws.com/download/GDC-PANCAN.basic_phenotype.tsv.gz\", sep=\"\\t\")\n",
    "    phe1 = phe1.loc[phe1.program == \"TCGA\", :].loc[:, ['sample', 'sample_type', 'project_id']].drop_duplicates(['sample'])\n",
    "    phe1['sample'] =  phe1.apply(lambda x : x['sample'][:-1], axis=1)\n",
    "    phe2 = pd.read_csv(\"https://tcga-pancan-atlas-hub.s3.us-east-1.amazonaws.com/download/TCGA_phenotype_denseDataOnlyDownload.tsv.gz\", sep=\"\\t\")\n",
    "    ph_join = pd.merge(left = phe2 , right = phe1, how = \"left\", on = \"sample\").dropna(subset=['project_id'])\n",
    "    \n",
    "    if cancer_type == \"PAN\" or cancer_type == \"PANCAN\":\n",
    "        filterd = ph_join.loc[ph_join['sample_type_y'] == \"Primary Tumor\", :]\n",
    "        sample_barcode = filterd[\"sample\"].tolist()\n",
    "    else:\n",
    "        filterd = ph_join.loc[(ph_join['sample_type_y'] == \"Primary Tumor\") & (ph_join['project_id'] == \"TCGA-\" + cancer_type) , :]\n",
    "        sample_barcode = filterd[\"sample\"].tolist()\n",
    "        \n",
    "    intersect_ = list(set(cols).intersection(sample_barcode))\n",
    "    \n",
    "    return intersect_\n",
    "\n",
    "def non_zero_column(DF):\n",
    "    sample_cnt = int(len(DF.columns) * 0.2)\n",
    "    zero_row = dict(DF.isin([0]).sum(axis=1))\n",
    "    non_remove_feature = list()\n",
    "\n",
    "    for key, value in zero_row.items():\n",
    "        if value < sample_cnt:\n",
    "            non_remove_feature.append(key)\n",
    "    \n",
    "    return non_remove_feature\n",
    "\n",
    "\n",
    "def determine_outlier_thresholds_iqr(dataframe, col_name, th1=0.05, th3=0.95):\n",
    "    quartile1 = dataframe[col_name].quantile(th1)\n",
    "    quartile3 = dataframe[col_name].quantile(th3)\n",
    "    iqr = quartile3 - quartile1\n",
    "    upper_limit = quartile3 + 1.5 * iqr\n",
    "    lower_limit = quartile1 - 1.5 * iqr\n",
    "    return lower_limit, upper_limit\n",
    "\n",
    "def check_outliers_iqr(dataframe, col_name):\n",
    "    lower_limit, upper_limit = determine_outlier_thresholds_iqr(dataframe, col_name)\n",
    "    if dataframe[(dataframe[col_name] > upper_limit) | (dataframe[col_name] < lower_limit)].any(axis=None):\n",
    "        return True\n",
    "    else: \n",
    "        return False\n",
    "    \n",
    "def replace_with_thresholds_iqr(dataframe, cols, th1=0.05, th3=0.95, replace=False):\n",
    "    from tabulate import tabulate\n",
    "    data = []\n",
    "    for col_name in cols:\n",
    "        if col_name != 'Outcome':\n",
    "            outliers_ = check_outliers_iqr(dataframe,col_name)\n",
    "            count = None\n",
    "            lower_limit, upper_limit = determine_outlier_thresholds_iqr(dataframe, col_name, th1, th3)\n",
    "            if outliers_:\n",
    "                count = dataframe[(dataframe[col_name] > upper_limit) | (dataframe[col_name] < lower_limit)][col_name].count()\n",
    "                if replace: \n",
    "                    dataframe.loc[(dataframe[col_name] < lower_limit), col_name] = lower_limit\n",
    "                    dataframe.loc[(dataframe[col_name] > upper_limit), col_name] = upper_limit\n",
    "            outliers_status = check_outliers_iqr(dataframe, col_name)\n",
    "            data.append([outliers_, outliers_status, count, col_name, lower_limit, upper_limit ])\n",
    "    table = tabulate(data, headers=['Outliers (Previously)', 'Outliers', 'Count', 'Column', 'Lower Limit', 'Upper Limit'], tablefmt='rst', numalign='right')\n",
    "   \n",
    "    return dataframe\n",
    "\n",
    "\n",
    "def get_dataset(path, cancer_type, std):\n",
    "    # non_scale dataset\n",
    "    # load raw DF\n",
    "    DF = pd.read_csv(path + \"/feature/input_feature_non_scale/\" + cancer_type + \".txt\", sep = \"\\t\")\n",
    "    pr_sample_DF = DF.iloc[:, 0:1]\n",
    "\n",
    "    if cancer_type == \"PANCAN\":\n",
    "        DF = DF.drop([\"Sample_barcode\"], axis=1)\n",
    "    else :\n",
    "        DF = DF.drop([\"Project\", \"Sample_barcode\"], axis=1)\n",
    "        \n",
    "    # remove outlier\n",
    "    DF = replace_with_thresholds_iqr(dataframe=DF, cols=DF.columns)\n",
    "    max_number_of_nas = DF.shape[0] / 2\n",
    "    DF = DF.loc[:, (DF.isnull().sum(axis=0) <= max_number_of_nas)]\n",
    "\n",
    "    # normalization\n",
    "    if std:\n",
    "        scalerX = StandardScaler()\n",
    "        scalerX.fit(DF)\n",
    "        DF_sacle = scalerX.transform(DF)\n",
    "    else :\n",
    "        scalerX = MinMaxScaler()\n",
    "        scalerX.fit(X_train)\n",
    "        DF_sacle = scalerX.transform(DF)\n",
    "    \n",
    "    # missing impute\n",
    "    imputer = KNNImputer(n_neighbors=10)\n",
    "    DF_sacle_impute = imputer.fit_transform(DF_sacle)\n",
    "\n",
    "    DF_sacle_impute = pd.DataFrame(DF_sacle_impute, columns=DF.columns)\n",
    "\n",
    "    return DF_sacle_impute\n",
    "\n",
    "# z = z_mean + sqrt(var) * epsilon\n",
    "def sampling(args):\n",
    "    \"\"\"Reparameterization trick by sampling from an isotropic unit Gaussian.\n",
    "    # Arguments\n",
    "        args (tensor): mean and log of variance of Q(z|X)\n",
    "    # Returns\n",
    "        z (tensor): sampled latent vector\n",
    "    \"\"\"\n",
    "\n",
    "    z_mean, z_log_var = args\n",
    "    batch = K.shape(z_mean)[0]\n",
    "    dim = K.int_shape(z_mean)[1]\n",
    "    # by default, random_normal has mean = 0 and std = 1.0\n",
    "    epsilon = K.random_normal(shape=(batch, dim))\n",
    "    return z_mean + K.exp(0.5 * z_log_var) * epsilon\n",
    "\n",
    "def recall_m(y_true, y_pred):\n",
    "    true_positives = K.sum(K.round(K.clip(y_true * y_pred, 0, 1)))\n",
    "    possible_positives = K.sum(K.round(K.clip(y_true, 0, 1)))\n",
    "    recall = true_positives / (possible_positives + K.epsilon())\n",
    "    return recall\n",
    "\n",
    "def precision_m(y_true, y_pred):\n",
    "    true_positives = K.sum(K.round(K.clip(y_true * y_pred, 0, 1)))\n",
    "    predicted_positives = K.sum(K.round(K.clip(y_pred, 0, 1)))\n",
    "    precision = true_positives / (predicted_positives + K.epsilon())\n",
    "    return precision\n",
    "\n",
    "def f1_score(y_true, y_pred):\n",
    "    precision = precision_m(y_true, y_pred)\n",
    "    recall = recall_m(y_true, y_pred)\n",
    "    return 2*((precision*recall)/(precision+recall+K.epsilon()))\n",
    "\n",
    "def get_train_test_DF(x, y):\n",
    "    X_train, X_test, y_train, y_test = train_test_split(x, y, test_size = 0.3, shuffle=True, random_state=32, stratify=y)\n",
    "    \n",
    "    return X_train, X_test, y_train, y_test\n",
    "\n",
    "def createFolder(path):\n",
    "    try:\n",
    "        if not os.path.exists(path):\n",
    "            os.makedirs(path)\n",
    "    except OSError:\n",
    "        print ('Folder already exists. ' +  path)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **Preprocessing**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **File Path**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/home/km/gitworking/Multi-omics-intergration/src'"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.getcwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "RAW_file_path = \"/home/km/gitworking/Multi-omics-intergration/RAW_DATA/\"\n",
    "PKL_file_path = \"/home/km/gitworking/Multi-omics-intergration/pkl/\"\n",
    "MODEL_path = \"/home/km/gitworking/Multi-omics-intergration/models/\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* **RNA gene expression**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [],
   "source": [
    "col = pd.read_csv(RAW_file_path + \"tcga_RSEM_Hugo_norm_count.gz\",\n",
    "                 sep = \"\\t\", index_col=0, nrows=0).columns.to_list()\n",
    "use_col = ['sample'] + cancer_select(cols=col, cancer_type=\"LIHC\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_chunk = pd.read_csv(RAW_file_path + \"tcga_RSEM_Hugo_norm_count.gz\",\n",
    "                 sep = \"\\t\", index_col=0, iterator=True, chunksize=50000, usecols=use_col)\n",
    "rna = pd.concat([chunk for chunk in df_chunk])\n",
    "rna = rna[rna.index.isin(non_zero_column(rna))].T"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* **miRNA**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [],
   "source": [
    "col = pd.read_csv(RAW_file_path + \"pancanMiRs_EBadjOnProtocolPlatformWithoutRepsWithUnCorrectMiRs_08_04_16.xena.gz\",\n",
    "                 sep = \"\\t\", index_col=0, nrows=0).columns.to_list()\n",
    "use_col = ['sample'] + cancer_select(cols=col, cancer_type=\"LIHC\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_chunk = pd.read_csv(RAW_file_path + \"pancanMiRs_EBadjOnProtocolPlatformWithoutRepsWithUnCorrectMiRs_08_04_16.xena.gz\",\n",
    "                 sep = \"\\t\", index_col=0, iterator=True, chunksize=50000, usecols=use_col)\n",
    "mirna = pd.concat([chunk for chunk in df_chunk])\n",
    "mirna = mirna[mirna.index.isin(non_zero_column(mirna))].T"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* **Methylation**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "col = pd.read_csv(RAW_file_path + \"jhu-usc.edu_PANCAN_HumanMethylation450.betaValue_whitelisted.tsv.synapse_download_5096262.xena.gz\",\n",
    "                 sep = \"\\t\", index_col=0, nrows=0).columns.to_list()\n",
    "use_col = ['sample'] + cancer_select(cols=col, cancer_type=\"LIHC\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_chunk = pd.read_csv(RAW_file_path + \"jhu-usc.edu_PANCAN_HumanMethylation450.betaValue_whitelisted.tsv.synapse_download_5096262.xena.gz\",\n",
    "                 sep = \"\\t\", index_col=0, iterator=True, chunksize=50000, usecols=use_col)\n",
    "mt = pd.concat([chunk for chunk in df_chunk])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "mt_map = pd.read_csv(RAW_file_path + \"probeMap_illuminaMethyl450_hg19_GPL16304_TCGAlegacy\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [],
   "source": [
    "mt_join = pd.merge(mt, mt_map, how = \"left\", left_on = \"sample\", right_on = \"#id\")\\\n",
    "         .drop(['chrom', 'chromStart', 'chromEnd', 'strand', '#id'], axis=1)\n",
    "mt_join = mt_join[mt_join.gene != \".\"]\n",
    "mt_join.dropna(subset = [\"gene\"], inplace=True)\n",
    "\n",
    "# gene mean \n",
    "mt_join_gene_filter = mt_join.groupby(['gene']).mean()\n",
    "mt_join_gene_filter = mt_join_gene_filter[mt_join_gene_filter.index.isin(non_zero_column(mt_join_gene_filter))].T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [],
   "source": [
    "rna['sample'] = rna.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [],
   "source": [
    "mirna['sample'] = mirna.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [],
   "source": [
    "mt_join_gene_filter['sample'] = mt_join_gene_filter.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {},
   "outputs": [],
   "source": [
    "merge_list = [rna, mirna, mt_join_gene_filter]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [],
   "source": [
    "omics = reduce(lambda left, right : pd.merge(left, right, on = \"sample\"), merge_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [],
   "source": [
    "omics.set_index('sample', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>RP11-16E12.2_x</th>\n",
       "      <th>RP11-181G12.2</th>\n",
       "      <th>CPHL1P_x</th>\n",
       "      <th>TCOF1_x</th>\n",
       "      <th>NSRP1_x</th>\n",
       "      <th>RP5-1000K24.2</th>\n",
       "      <th>SPPL3_x</th>\n",
       "      <th>AF001548.5</th>\n",
       "      <th>OPA3_x</th>\n",
       "      <th>OPA1_x</th>\n",
       "      <th>...</th>\n",
       "      <th>chr22-38_28785274-29006793.1,ZNRF3-AS1,ZNRF3</th>\n",
       "      <th>pRNA</th>\n",
       "      <th>pRNA,AP001347.6</th>\n",
       "      <th>pk_y</th>\n",
       "      <th>pk,MLK7-AS1</th>\n",
       "      <th>snoR1,SNHG19,SNORD60,TRAF7</th>\n",
       "      <th>snoU13,KIF9</th>\n",
       "      <th>snoU2_19,EIF1AX,snoU2-30</th>\n",
       "      <th>uc_338,SMARCA4</th>\n",
       "      <th>yR211F11.2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-CC-A3MA-01</th>\n",
       "      <td>3.0170</td>\n",
       "      <td>4.7785</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>11.5952</td>\n",
       "      <td>10.4224</td>\n",
       "      <td>3.6565</td>\n",
       "      <td>10.3479</td>\n",
       "      <td>5.6826</td>\n",
       "      <td>9.9501</td>\n",
       "      <td>10.5289</td>\n",
       "      <td>...</td>\n",
       "      <td>0.226746</td>\n",
       "      <td>0.253931</td>\n",
       "      <td>0.802173</td>\n",
       "      <td>0.262715</td>\n",
       "      <td>0.841819</td>\n",
       "      <td>0.040540</td>\n",
       "      <td>0.901541</td>\n",
       "      <td>0.877858</td>\n",
       "      <td>0.951853</td>\n",
       "      <td>0.918813</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-DD-AAW0-01</th>\n",
       "      <td>2.1896</td>\n",
       "      <td>2.6651</td>\n",
       "      <td>2.8547</td>\n",
       "      <td>10.4419</td>\n",
       "      <td>9.7246</td>\n",
       "      <td>3.4323</td>\n",
       "      <td>10.1216</td>\n",
       "      <td>4.5154</td>\n",
       "      <td>10.0724</td>\n",
       "      <td>11.9185</td>\n",
       "      <td>...</td>\n",
       "      <td>0.281292</td>\n",
       "      <td>0.353733</td>\n",
       "      <td>0.934289</td>\n",
       "      <td>0.274041</td>\n",
       "      <td>0.767279</td>\n",
       "      <td>0.035665</td>\n",
       "      <td>0.236923</td>\n",
       "      <td>0.429889</td>\n",
       "      <td>0.905745</td>\n",
       "      <td>0.931620</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-DD-AACA-01</th>\n",
       "      <td>3.2386</td>\n",
       "      <td>2.1751</td>\n",
       "      <td>0.7682</td>\n",
       "      <td>10.7915</td>\n",
       "      <td>10.3504</td>\n",
       "      <td>2.8736</td>\n",
       "      <td>10.0403</td>\n",
       "      <td>4.0645</td>\n",
       "      <td>9.6837</td>\n",
       "      <td>11.4174</td>\n",
       "      <td>...</td>\n",
       "      <td>0.245278</td>\n",
       "      <td>0.149060</td>\n",
       "      <td>0.506923</td>\n",
       "      <td>0.247678</td>\n",
       "      <td>0.726268</td>\n",
       "      <td>0.032140</td>\n",
       "      <td>0.235719</td>\n",
       "      <td>0.392821</td>\n",
       "      <td>0.932847</td>\n",
       "      <td>0.312001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-K7-AAU7-01</th>\n",
       "      <td>5.5060</td>\n",
       "      <td>5.9382</td>\n",
       "      <td>7.3385</td>\n",
       "      <td>11.0277</td>\n",
       "      <td>10.0970</td>\n",
       "      <td>3.3956</td>\n",
       "      <td>11.3990</td>\n",
       "      <td>5.0913</td>\n",
       "      <td>10.6333</td>\n",
       "      <td>11.1081</td>\n",
       "      <td>...</td>\n",
       "      <td>0.255657</td>\n",
       "      <td>0.300331</td>\n",
       "      <td>0.639749</td>\n",
       "      <td>0.258850</td>\n",
       "      <td>0.776524</td>\n",
       "      <td>0.044826</td>\n",
       "      <td>0.120922</td>\n",
       "      <td>0.850816</td>\n",
       "      <td>0.929263</td>\n",
       "      <td>0.828439</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-RC-A7SH-01</th>\n",
       "      <td>6.6832</td>\n",
       "      <td>4.9831</td>\n",
       "      <td>5.7378</td>\n",
       "      <td>10.8955</td>\n",
       "      <td>9.0594</td>\n",
       "      <td>0.9913</td>\n",
       "      <td>11.3930</td>\n",
       "      <td>6.7918</td>\n",
       "      <td>10.7746</td>\n",
       "      <td>10.6830</td>\n",
       "      <td>...</td>\n",
       "      <td>0.258233</td>\n",
       "      <td>0.408436</td>\n",
       "      <td>0.924842</td>\n",
       "      <td>0.246913</td>\n",
       "      <td>0.794186</td>\n",
       "      <td>0.046974</td>\n",
       "      <td>0.198767</td>\n",
       "      <td>0.889381</td>\n",
       "      <td>0.901486</td>\n",
       "      <td>0.876262</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-2Y-A9H8-01</th>\n",
       "      <td>4.2129</td>\n",
       "      <td>6.1754</td>\n",
       "      <td>3.5807</td>\n",
       "      <td>10.4468</td>\n",
       "      <td>9.1961</td>\n",
       "      <td>2.6966</td>\n",
       "      <td>10.4525</td>\n",
       "      <td>3.5117</td>\n",
       "      <td>10.1593</td>\n",
       "      <td>11.2977</td>\n",
       "      <td>...</td>\n",
       "      <td>0.338572</td>\n",
       "      <td>0.376402</td>\n",
       "      <td>0.962969</td>\n",
       "      <td>0.290296</td>\n",
       "      <td>0.783476</td>\n",
       "      <td>0.034968</td>\n",
       "      <td>0.840098</td>\n",
       "      <td>0.825663</td>\n",
       "      <td>0.923310</td>\n",
       "      <td>0.893791</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-2Y-A9GT-01</th>\n",
       "      <td>4.8918</td>\n",
       "      <td>5.7623</td>\n",
       "      <td>1.2425</td>\n",
       "      <td>9.7209</td>\n",
       "      <td>9.1073</td>\n",
       "      <td>3.8739</td>\n",
       "      <td>10.4595</td>\n",
       "      <td>5.1868</td>\n",
       "      <td>10.3863</td>\n",
       "      <td>10.7661</td>\n",
       "      <td>...</td>\n",
       "      <td>0.313733</td>\n",
       "      <td>0.369372</td>\n",
       "      <td>0.546060</td>\n",
       "      <td>0.311700</td>\n",
       "      <td>0.807100</td>\n",
       "      <td>0.045246</td>\n",
       "      <td>0.182878</td>\n",
       "      <td>0.939182</td>\n",
       "      <td>0.899394</td>\n",
       "      <td>0.609828</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-DD-A73C-01</th>\n",
       "      <td>0.9754</td>\n",
       "      <td>3.7613</td>\n",
       "      <td>1.5521</td>\n",
       "      <td>9.8084</td>\n",
       "      <td>9.8115</td>\n",
       "      <td>1.5521</td>\n",
       "      <td>10.4740</td>\n",
       "      <td>3.0522</td>\n",
       "      <td>10.4710</td>\n",
       "      <td>12.4116</td>\n",
       "      <td>...</td>\n",
       "      <td>0.355588</td>\n",
       "      <td>0.377962</td>\n",
       "      <td>0.930226</td>\n",
       "      <td>0.269236</td>\n",
       "      <td>0.841078</td>\n",
       "      <td>0.036144</td>\n",
       "      <td>0.067261</td>\n",
       "      <td>0.942604</td>\n",
       "      <td>0.923079</td>\n",
       "      <td>0.798672</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-2Y-A9H2-01</th>\n",
       "      <td>3.9896</td>\n",
       "      <td>4.6428</td>\n",
       "      <td>2.1070</td>\n",
       "      <td>11.3878</td>\n",
       "      <td>10.5246</td>\n",
       "      <td>3.6528</td>\n",
       "      <td>10.3017</td>\n",
       "      <td>4.9485</td>\n",
       "      <td>10.8637</td>\n",
       "      <td>11.1011</td>\n",
       "      <td>...</td>\n",
       "      <td>0.282299</td>\n",
       "      <td>0.493378</td>\n",
       "      <td>0.949594</td>\n",
       "      <td>0.265444</td>\n",
       "      <td>0.790375</td>\n",
       "      <td>0.047294</td>\n",
       "      <td>0.404722</td>\n",
       "      <td>0.916887</td>\n",
       "      <td>0.930275</td>\n",
       "      <td>0.862763</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-DD-A115-01</th>\n",
       "      <td>5.2355</td>\n",
       "      <td>4.8019</td>\n",
       "      <td>3.7252</td>\n",
       "      <td>10.1904</td>\n",
       "      <td>9.4643</td>\n",
       "      <td>1.1522</td>\n",
       "      <td>10.4583</td>\n",
       "      <td>4.5729</td>\n",
       "      <td>10.5064</td>\n",
       "      <td>10.9770</td>\n",
       "      <td>...</td>\n",
       "      <td>0.312623</td>\n",
       "      <td>0.325445</td>\n",
       "      <td>0.775010</td>\n",
       "      <td>0.282641</td>\n",
       "      <td>0.788807</td>\n",
       "      <td>0.039779</td>\n",
       "      <td>0.090617</td>\n",
       "      <td>0.911030</td>\n",
       "      <td>0.941144</td>\n",
       "      <td>0.529207</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>362 rows × 58419 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 RP11-16E12.2_x  RP11-181G12.2  CPHL1P_x  TCOF1_x  NSRP1_x  \\\n",
       "sample                                                                       \n",
       "TCGA-CC-A3MA-01          3.0170         4.7785    0.0000  11.5952  10.4224   \n",
       "TCGA-DD-AAW0-01          2.1896         2.6651    2.8547  10.4419   9.7246   \n",
       "TCGA-DD-AACA-01          3.2386         2.1751    0.7682  10.7915  10.3504   \n",
       "TCGA-K7-AAU7-01          5.5060         5.9382    7.3385  11.0277  10.0970   \n",
       "TCGA-RC-A7SH-01          6.6832         4.9831    5.7378  10.8955   9.0594   \n",
       "...                         ...            ...       ...      ...      ...   \n",
       "TCGA-2Y-A9H8-01          4.2129         6.1754    3.5807  10.4468   9.1961   \n",
       "TCGA-2Y-A9GT-01          4.8918         5.7623    1.2425   9.7209   9.1073   \n",
       "TCGA-DD-A73C-01          0.9754         3.7613    1.5521   9.8084   9.8115   \n",
       "TCGA-2Y-A9H2-01          3.9896         4.6428    2.1070  11.3878  10.5246   \n",
       "TCGA-DD-A115-01          5.2355         4.8019    3.7252  10.1904   9.4643   \n",
       "\n",
       "                 RP5-1000K24.2  SPPL3_x  AF001548.5   OPA3_x   OPA1_x  ...  \\\n",
       "sample                                                                 ...   \n",
       "TCGA-CC-A3MA-01         3.6565  10.3479      5.6826   9.9501  10.5289  ...   \n",
       "TCGA-DD-AAW0-01         3.4323  10.1216      4.5154  10.0724  11.9185  ...   \n",
       "TCGA-DD-AACA-01         2.8736  10.0403      4.0645   9.6837  11.4174  ...   \n",
       "TCGA-K7-AAU7-01         3.3956  11.3990      5.0913  10.6333  11.1081  ...   \n",
       "TCGA-RC-A7SH-01         0.9913  11.3930      6.7918  10.7746  10.6830  ...   \n",
       "...                        ...      ...         ...      ...      ...  ...   \n",
       "TCGA-2Y-A9H8-01         2.6966  10.4525      3.5117  10.1593  11.2977  ...   \n",
       "TCGA-2Y-A9GT-01         3.8739  10.4595      5.1868  10.3863  10.7661  ...   \n",
       "TCGA-DD-A73C-01         1.5521  10.4740      3.0522  10.4710  12.4116  ...   \n",
       "TCGA-2Y-A9H2-01         3.6528  10.3017      4.9485  10.8637  11.1011  ...   \n",
       "TCGA-DD-A115-01         1.1522  10.4583      4.5729  10.5064  10.9770  ...   \n",
       "\n",
       "                 chr22-38_28785274-29006793.1,ZNRF3-AS1,ZNRF3      pRNA  \\\n",
       "sample                                                                    \n",
       "TCGA-CC-A3MA-01                                      0.226746  0.253931   \n",
       "TCGA-DD-AAW0-01                                      0.281292  0.353733   \n",
       "TCGA-DD-AACA-01                                      0.245278  0.149060   \n",
       "TCGA-K7-AAU7-01                                      0.255657  0.300331   \n",
       "TCGA-RC-A7SH-01                                      0.258233  0.408436   \n",
       "...                                                       ...       ...   \n",
       "TCGA-2Y-A9H8-01                                      0.338572  0.376402   \n",
       "TCGA-2Y-A9GT-01                                      0.313733  0.369372   \n",
       "TCGA-DD-A73C-01                                      0.355588  0.377962   \n",
       "TCGA-2Y-A9H2-01                                      0.282299  0.493378   \n",
       "TCGA-DD-A115-01                                      0.312623  0.325445   \n",
       "\n",
       "                 pRNA,AP001347.6      pk_y  pk,MLK7-AS1  \\\n",
       "sample                                                    \n",
       "TCGA-CC-A3MA-01         0.802173  0.262715     0.841819   \n",
       "TCGA-DD-AAW0-01         0.934289  0.274041     0.767279   \n",
       "TCGA-DD-AACA-01         0.506923  0.247678     0.726268   \n",
       "TCGA-K7-AAU7-01         0.639749  0.258850     0.776524   \n",
       "TCGA-RC-A7SH-01         0.924842  0.246913     0.794186   \n",
       "...                          ...       ...          ...   \n",
       "TCGA-2Y-A9H8-01         0.962969  0.290296     0.783476   \n",
       "TCGA-2Y-A9GT-01         0.546060  0.311700     0.807100   \n",
       "TCGA-DD-A73C-01         0.930226  0.269236     0.841078   \n",
       "TCGA-2Y-A9H2-01         0.949594  0.265444     0.790375   \n",
       "TCGA-DD-A115-01         0.775010  0.282641     0.788807   \n",
       "\n",
       "                 snoR1,SNHG19,SNORD60,TRAF7  snoU13,KIF9  \\\n",
       "sample                                                     \n",
       "TCGA-CC-A3MA-01                    0.040540     0.901541   \n",
       "TCGA-DD-AAW0-01                    0.035665     0.236923   \n",
       "TCGA-DD-AACA-01                    0.032140     0.235719   \n",
       "TCGA-K7-AAU7-01                    0.044826     0.120922   \n",
       "TCGA-RC-A7SH-01                    0.046974     0.198767   \n",
       "...                                     ...          ...   \n",
       "TCGA-2Y-A9H8-01                    0.034968     0.840098   \n",
       "TCGA-2Y-A9GT-01                    0.045246     0.182878   \n",
       "TCGA-DD-A73C-01                    0.036144     0.067261   \n",
       "TCGA-2Y-A9H2-01                    0.047294     0.404722   \n",
       "TCGA-DD-A115-01                    0.039779     0.090617   \n",
       "\n",
       "                 snoU2_19,EIF1AX,snoU2-30  uc_338,SMARCA4  yR211F11.2  \n",
       "sample                                                                 \n",
       "TCGA-CC-A3MA-01                  0.877858        0.951853    0.918813  \n",
       "TCGA-DD-AAW0-01                  0.429889        0.905745    0.931620  \n",
       "TCGA-DD-AACA-01                  0.392821        0.932847    0.312001  \n",
       "TCGA-K7-AAU7-01                  0.850816        0.929263    0.828439  \n",
       "TCGA-RC-A7SH-01                  0.889381        0.901486    0.876262  \n",
       "...                                   ...             ...         ...  \n",
       "TCGA-2Y-A9H8-01                  0.825663        0.923310    0.893791  \n",
       "TCGA-2Y-A9GT-01                  0.939182        0.899394    0.609828  \n",
       "TCGA-DD-A73C-01                  0.942604        0.923079    0.798672  \n",
       "TCGA-2Y-A9H2-01                  0.916887        0.930275    0.862763  \n",
       "TCGA-DD-A115-01                  0.911030        0.941144    0.529207  \n",
       "\n",
       "[362 rows x 58419 columns]"
      ]
     },
     "execution_count": 140,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "omics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "multiomic",
   "language": "python",
   "name": "multiomics"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
